### PR TITLE
Review Notebook 03

### PR DESCRIPTION
**Pull request by beneyal**
1. The "Input and output shapes" image shows NHWC. Might be worthwhile to change to NCHW.
2. In addition to being more computationally efficient, mini-batching also gives the model more opportunities to improve. Batch Gradient Descent would calculate the gradient once per epoch and that's it. Mini-batch SGD gives you `dataset_size // batch_size` weight updates.
3. After reading [this](https://pytorch.org/docs/stable/notes/randomness.html), I'm not 100% sure that `torch.manual_seed(42)` will give the same results for everybody.
4. I understood that model 0 will be ran on both CPU and GPU, but it is only run on CPU.
5. Image above 7.2 also uses NHWC instead of NCHW
6. While it is possible to type hint the `accuracy_fn` parameter, I decided against it, as this is a "noisy" type hint
7. I'm running on PyTorch 1.11 and apparently, running `conv_layer(test_image)` actually works without a batch dimension and returns [10, 62, 62] shape result
8. Added type hints where missing
9. Changed the model vs accuracy bar chart to be horizontal, that way the model names are more readable 🙂

### CODE REVIEW

**Comment by mrdbourke** [comment]
Thank you!

Going to go through these today.

**Comment by mrdbourke** [comment]
> 1. The "Input and output shapes" image shows NHWC. Might be worthwhile to change to NCHW.

Done.

> 2. In addition to being more computationally efficient, mini-batching also gives the model more opportunities to improve. Batch Gradient Descent would calculate the gradient once per epoch and that's it. Mini-batch SGD gives you `dataset_size // batch_size` weight updates.

Good tip! Included this:

> It also gives your model more opportunities to improve.
>
> With **mini-batches** (small portions of the data), gradient descent is performed more often per epoch (once per mini-batch rather than once per epoch).

> 3. After reading [this](https://pytorch.org/docs/stable/notes/randomness.html), I'm not 100% sure that `torch.manual_seed(42)` will give the same results for everybody.

You're right. I've hinted at this throughout the videos.

Randomness shouldn't be expected to be 100% deterministic, even with setting seeds but the *direction* (e.g. lower loss etc) should be similar.

The numbers might not be *exactly* the same but the improvements should be similar.

> 4. I understood that model 0 will be ran on both CPU and GPU, but it is only run on CPU.

Added a note to make this clearer: 
* `model_0` - CPU (hard-coded)
* `model_1` - GPU (device-agnostic)
* `model_2` - GPU (`device-agnostic)

> 5. Image above 7.2 also uses NHWC instead of NCHW

Fixed!

> 6. While it is possible to type hint the `accuracy_fn` parameter, I decided against it, as this is a "noisy" type hint

I had similar thoughts here.

> 7. I'm running on PyTorch 1.11 and apparently, running `conv_layer(test_image)` actually works without a batch dimension and returns [10, 62, 62] shape result

Added a note for this, thank you.

> 8. Added type hints where missing

Thank you!

> 9. Changed the model vs accuracy bar chart to be horizontal, that way the model names are more readable 🙂 

Looks better!

See updated version: https://github.com/mrdbourke/pytorch-deep-learning/blob/main/03_pytorch_computer_vision.ipynb
